core/common/src/Sinks.kt

```diff
@@ -135,6 +107,34 @@ public fun Sink.writeDecimalLong(long: Long) {
     }
 }
 
+private fun countDigitsIn(v: Long): Int {
+    val guess = ((64 - v.countLeadingZeroBits()) * 10) ushr 5
+    return guess + (if (v > DigitCountToLargestValue[guess]) 1 else 0)
```

Comment:
IIRC from the time I read Romain's blogpost, by extending `DigitCountToLargestValue`'s length to next power of two (32 in this case) and replacing `DigitCountToLargestValue[guess]` with `DigitCountToLargestValue[guess.and(0x1f)]` you can win a few extra percents of performance on JVM (as it should optimize out bounds checks performed on array access).

Reply:
`DigitCountToLargestValue` is actually slightly different than the table used in the blogpost:

```kotlin
private val PowersOfTen = longArrayOf(
    0,
    10,
    100,
    1000,
    10000,
    100000,
    1000000,
    10000000,
    100000000,
    1000000000,
    10000000000,
    100000000000,
    1000000000000,
    10000000000000,
    100000000000000,
    1000000000000000,
    10000000000000000,
    100000000000000000,
    1000000000000000000
)
```

The main reason is that the original table doesn't work when the input is `Long.MAX_VALUE`, as it's bigger than 10^18 (last value in the array), but 10^19 is outside of the Long range.

I wonder if the one in the PR performs better? Worth benchmarking them against each other?

Reply:
What I meant is that loads from `DigitCountToLargestValue` table are compiled into a code that checks if an index is within array's bounds before performing a load.
However, if compiler can prove that indices are always in bounds, it'll abstain from generating the check. 
By expanding the table to have a power-of-two length (and filling meaningless cells with, let's say, `-1`) and then explicitly truncating index's most significant bits (i.e., dividing an index by table's length and taking the remainder), we can hint a compiler that a value is always in bounds and it'll generate _faster_ code: https://gist.github.com/fzhinkin/42997a2cfc18a437f88e9c31bef969c9

Reply:
BTW I checked and on Android the power-of-two array + truncation doesn't remove the bounds check. It just adds an extra instruction. See https://godbolt.org/z/jdTzMcxbf